### ISSUE TITLE
Some data is missing on recent queries, but re-appears during longer range queries

### ISSUE THREAD

**Issue by johnseekins** [VictoriaMetrics/VictoriaMetrics#2798]
**Describe the bug**

Seeing something very strange in queries to VictoriaMetrics this morning:

```
john@framework:~$ date "+%s" && curl -H 'Authorization: Bearer <>' -X POST -Ss https://metrics/api/v1/query -d 'query=github_branches_total[48h]' | jq .data.result[].values[-1]
1656428452
[
  1656425771,
  "367"
]
john@framework:~$ date "+%s" && curl -H 'Authorization: Bearer <>' -X POST -Ss https://metrics/api/v1/query -d 'query=github_branches_total[24h]' | jq .data.result[].values[-1]
1656428456
```

When I do a larger raw query, I get data points from within the last hour (as expected). Shorter ranges? No data.
(Using raw exports, I see the raw data, too)

```
john@framework:~$ date "+%s" && curl -H 'Authorization: Bearer <>' -X POST -Ss https://metrics/api/v1/export -d 'match[]=github_branches_total' | jq '"\(.values[-1]) \(.timestamps[-1])"'
1656428706
"367 1656425771000"
```

So this implies (to me) that something is weird in the query function. Or perhaps some cache is corrupted? I have tried restarting the instance and it seems to quickly start failing again.

What's particularly interesting is not _all_ data is affected:

```
date "+%s" && curl -H 'Authorization: Bearer <>' -X POST -Ss https://metrics/api/v1/query -d 'query=github_branch_window_commits_total[2h]' | jq .data.result[].values[-1]
1656432562
[
  1656425771,
  "1"
]
[
  1656429413,
  "1"
]
```

Also, adding `nocache=1` to the query string doesn't seem to help.

**Expected behavior**
Shorter range queries also result in the same data that _clearly_ exists in the data store.

**Logs**
Only logs that may be relevant are rollupCache reset messages.

**Version**
1.78.0

**Used command-line flags**
```
-retentionPeriod 3y -selfScrapeInterval 1m -promscrape.config /scrape.yml -http.disableResponseCompression -search.cacheTimestampOffset=30m0s
```

**Comment by johnseekins** [comment]
Interesting additional notes:

1. It now takes going back `[3d]` for the data to populate properly.
2. This may be related to these changes (according to @f41gh7 ): https://github.com/VictoriaMetrics/VictoriaMetrics/commit/55e7afae3a0c20e53e790e6c6afb02fff848f868#diff-31747331b2a09f4e4df81c672333af155957a49dfcfec9ac3834b29ac78aecaa

**Comment by thislyc** [comment]
I feel like I have the same problem

**Comment by hagen1778** [comment]
Hi @johnseekins!

Could you please try following things:
1. Try exporting data via /api/v1/export with params `start=-2d` and `start=-1d`? If the recent values will be different than it is likely an issue with daily-index. By default, /api/v1/export uses global index.
2. If p.1 didn't show the issue, try checking if returned response isn't partial. Check if `-replicationFactor` is set on vmselect and remove it. Or try sending `deny_partial_response=1` param via GET /api/v1/query.
3. If no issue found yet, try executing queries for 24h and 48h with [tracing enabled ](https://docs.victoriametrics.com/#query-tracing) and compare the output.

**Comment by johnseekins** [comment]
>     2. If p.1 didn't show the issue, try checking if returned response isn't partial. Check if `-replicationFactor` is set on vmselect and remove it. Or try sending `deny_partial_response=1` param via GET /api/v1/query.

I can answer this quickly: I'm not using the cluster version. ;)

**Comment by johnseekins** [comment]
Interestingly, over the last few days, the time I have to scan back to catch failing stats has grown:

```
john@framework:~$ date "+%s" && curl -H 'Authorization: Bearer <>' -X POST -Ss https://metrics/api/v1/export -d 'match[]=github_weekly_commits_total' -d "start=-2d" | jq '"\(.values[-1]) \(.timestamps[-1])"'
1656598752
john@framework:~$ date "+%s" && curl -H 'Authorization: Bearer <>' -X POST -Ss https://metricsapi/v1/export -d 'match[]=github_weekly_commits_total' -d "start=-4d" | jq '"\(.values[-1]) \(.timestamps[-1])"'
1656598757
"3 1656595445000"
```

Also interestingly? Today, some of the queries that _had_ been failing now aren't. Which is why I'm using a different stat today.

**Comment by little-liang** [comment]
i have a same problem

**Comment by johnseekins** [comment]
I attempted downgrading to 1.77.2 while a fix is ongoing for this and ran into an interesting problem:
The caches were still corrupted, so I still couldn't load recent data for some stats. I think anyone who upgraded to 1.78.0 will need a way to safely clean their caches to fully fix this problem.

**Comment by hagen1778** [comment]
Hi @johnseekins! Try requesting wider time range (more than 24h) or [deleting the caches](https://docs.victoriametrics.com/?highlight=reset_cache_on_startup#cache-removal).

**Comment by f41gh7** [comment]
Yes, cache reset at vmsingle/vmstorage is needed if you upgrade from v1.78.0 to v1.78.1 (or v1.78.1-rc1). 

I've prepared docker images with fix for 1.78 release:
- victoriametrics/vmstorage:v1.78.1-rc1
- victoriametrics/victoria-metrics:v1.78.1-rc1

**Comment by Bingmang** [comment]
Meet the same problem. Data is missing when I query 5 minutes, but comes back when I query 6 hours.

**Comment by valyala** [comment]
> why the wider time range will do the trick? Could you explain it please.

VictoriaMetrics registers incoming time series in the inverted index for fast lookups of time series matching the given label filters (aka [series selectors](https://prometheus.io/docs/prometheus/latest/querying/basics/#time-series-selectors)). VictoriaMetrics maintains two inverted indexes in the `indexdb`:
- Global index across all the time series encountered during the configured [retention](https://docs.victoriametrics.com/#retention). This index is used for series lookup on long time ranges, which exceed 40 days.
- Per-day index, which contains entries only for time series with at least a single sample ingested at the particular day. This index is used for series lookup on relatively short time ranges (up to 40 days). This index improves series lookup speed over short time ranges when old series are substituted with new series at high rate (aka [high churn rate](https://docs.victoriametrics.com/FAQ.html#what-is-high-churn-rate)). For example, if 10M new series are created every day, then the global index would contain information about 10M*100=1B (one billion) series for `-retentionPeriod=100d` (e.g. 100 days retention), while every per-day index would contain information only for 10M series. The lookup performance over 10M series is better than the lookup performance over 1000M series.

The v1.78.0 contains a bug, which prevents from registering a small percentage of time series in per-day inverted index. This means that such time series are missing in query results over short time ranges. When the time range exceeds a day, then this increases chances that the series will be found in other per-day indexes.

The bug has been fixed in [VictoriaMetrics v1.78.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this issue as fixed.
